### PR TITLE
HttpCacheClient: make upload async

### LINKED ISSUE #6091: Jobs generating large amounts of files drastically slow down building with remote cache.
### Description of the problem / feature request:

Building the go standard library with remote caching enabled is considerably slower than building locally.

IRL discussion with @buchgr resulted in the conclusion that the contents of the standard library were uploaded serially within Bazel, which lead to major performance decreases. 
It was noted that some concurrency is present for the rest of the build, since Bazel executes multiple jobs at once, resulting in pseudo-concurrent results; but since the go standard library counts as a single job, the upload is then performed serially.

### Feature requests: what underlying problem are you trying to solve with this feature?

Decrease build time when populating empty remote caches by allowing concurrent uploads of files in individual actions.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Build any go project using an empty remote cache; on normal internet connections, this can take a _long_ time as the go standard library files are uploaded one by one.

### Example timings, running bazel test on one of our projects with a clean local cache:
With remote caching:
```
bazel test --remote_http_cache=https://storage.googleapis.com/<our cache>  
0.07s user 0.15s system 0% cpu 14:07.77 total
```
No remote caching:
```
bazel test
0.05s user 0.08s system 0% cpu 2:46.71 total
```

### PR DESCRIPTION
**Pull request by ulfjack**
This matches the behavior for download, and improves performance for actions
with a lot of outputs.

The code was already using futures in the API and in the implementation,
but the glue code blocked on the internal future and returned an immediate
future to the caller, rather than doing it async.

Fixes #6091.

Change-Id: I3151aa96b879323e0000d3209f6b9bc8be0066d4

### CODE REVIEW

**Comment by ulfjack** [comment]
@coeuvre

**Comment by coeuvre** [comment]
Thanks for the PR! LGTM. But some tests failed, any idea with that?

**Comment by ulfjack** [comment]
It was a bug in the change which I just fixed.

**Comment by ulfjack** [comment]
All tests pass!

**Review by coeuvre — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java [hunk 1]
```diff
@@ -285,64 +287,58 @@ public void channelCreated(Channel ch) {
...
+            .addListener(
+                    (Future<Channel> channelAcquired) -> {
+                      if (!channelAcquired.isSuccess()) {
+                        channelReady.setFailure(channelAcquired.cause());
+                        return;
+                      }
+
+                      try {
+                        Channel ch = channelAcquired.getNow();
+                        ChannelPipeline p = ch.pipeline();
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java:303]: nit: can you please change `p` to `pipeline` (or similar)?

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java [hunk 2]
```diff
@@ -588,85 +583,97 @@ public void close() {
...
-        }
-        in.close();
-        if (ch != null) {
-          releaseUploadChannel(ch);
-        }
-      }
+      SettableFuture<Void> result = SettableFuture.create();
+      acquireUploadChannel()
+          .addListener(
+              (Future<Channel> chP) -> {
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java:590]: nit: `channelPromise`?

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java [hunk 3]
```diff
@@ -588,85 +583,97 @@ public void close() {
...
+            });
+  }
+
   @Override
   public ListenableFuture<Void> uploadFile(Digest digest, Path file) {
-    try (InputStream in = file.getInputStream()) {
-      uploadBlocking(digest.getHash(), digest.getSizeBytes(), in, /* casUpload= */ true);
-    } catch (IOException | InterruptedException e) {
+    try {
+      return uploadAsync(digest.getHash(), digest.getSizeBytes(), file.getInputStream(), true);
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java:660]: nit: please re-add `/* casUpload= */`
(same a few lines below and line 695)

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java [hunk 4]
```diff
@@ -588,85 +583,97 @@ public void close() {
...
+                        (f) ->{
+                          releaseUploadChannel(ch);
+                          if (f.isSuccess()) {
+                            result.set(null);
+                          } else {
+                            Throwable cause = f.cause();
+                            if (cause instanceof HttpException) {
+                              HttpResponse response = ((HttpException) cause).response();
+                              try {
+                                // If the error is due to an expired auth token and we can reset the input stream,
```
Comment (meisterT) [inline src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java:608]: nit: line too long

**Review by meisterT — COMMENTED** [review]
Some minor nits

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java (hunk 1 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java]: Fixed. I changed the download side as well.

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java (hunk 2 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java]: Changed. I didn't change the download side.

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java (hunk 3 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java]: Sorry about that. Not sure how that got lost.

File: src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java (hunk 4 above)
Comment (ulfjack) [reply src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java]: Fixed. Also a couple more of these.

**Comment by ulfjack** [comment]
PTAL

**Review by meisterT — APPROVED** [review]

**Pull request closed without merging**